[clang][Driver] Add -fuse-lipo option

khyperia opened the pull request:
Partially fixes https://github.com/llvm/llvm-project/issues/59552 by adding a new option `-fuse-lipo` that can specify the tool name to be used by clang-driver for the lipo action. For example, pass `-fuse-lipo=llvm-lipo` to use `llvm-lipo` instead of the default `lipo`.

khyperia committed:
Add -fuse-lipo option

carlocab commented on clang/lib/Driver/ToolChains/Darwin.cpp:
```diff
@@ -910,7 +910,10 @@ void darwin::Lipo::ConstructJob(Compilation &C, const JobAction &JA,
   }
 
-  const char *Exec = Args.MakeArgString(getToolChain().GetProgramPath("lipo"));
+  std::string LipoName =
+      std::string(Args.getLastArgValue(options::OPT_fuse_lipo_EQ, "lipo"));
+  const char *Exec =
+      Args.MakeArgString(getToolChain().GetProgramPath(LipoName.c_str()));
```
Seems a bit much to create a throwaway `std::string` here. Something like this should work:
```suggestion
  StringRef LipoName =
      Args.getLastArgValue(options::OPT_fuse_lipo_EQ, "lipo");
  const char *Exec =
      Args.MakeArgString(getToolChain().GetProgramPath(LipoName.data()));
```

khyperia replied:
Ah, StringRef::data() is documented as "data - Get a pointer to the start of the string (which may not be null terminated)". Because we're using it as a null-terminated string, I thought that making a copy is necessary to ensure it's null terminated, in case getLastArgValue ever changes to not return a null-terminated string. Depending on it always returning a null terminated string seems like what a lot of other code does already, though, I'll make that change.

carlocab reviewed (commented):
Thanks for the PR! This change needs new tests to be added.

khyperia committed:
Add basic fuse-lipo.c tests

khyperia commented:
@carlocab I've removed the temporary std::string and added a guess at basic tests. Let me know if there's additional tests in particular that you're thinking of!

cachemeifyoucan commented on clang/test/Driver/fuse-lipo.c:
```diff
@@ -0,0 +1,11 @@
+// RUN: %clang %s -### --target=arm64-apple-darwin -arch x86_64 -arch arm64 -fuse-lipo=llvm-lipo 2>&1 | FileCheck -check-prefix=TEST1 %s
```
Nit: Add a test case to check when the flag is not supplied.

cachemeifyoucan reviewed (commented):
LGTM in general with a comment in test.

For discussion. Is it better if the option supplies the full path to lipo or just the name? Full path seems to be easy to use, but might deserve a warning if the tool doesn't exist.
If just the name, it might be better to rename the option to something like `-fuse-lipo-program=`.

khyperia committed:
Add fuse-lipo test without arg

khyperia commented:
> LGTM in general with a comment in test.
> 
> For discussion. Is it better if the option supplies the full path to lipo or just the name? Full path seems to be easy to use, but might deserve a warning if the tool doesn't exist. If just the name, it might be better to rename the option to something like `-fuse-lipo-program=`.

As a prefix: I am a new contributor, if you or someone else experienced has an opinion here I will gladly blindly follow it.

I tried to take precedence from `-fuse-ld` and `--ld-path`, so `-fuse-lipo` takes a name, and a hypothetical future `--lipo-path` would be a full path. Perhaps these are only named this way due to legacy compatibility though, and new naming conventions should be something else, `fuse-lipo-program` as you say? I don't know the history and context here. Let me know what I should do!

> Nit: Add a test case to check when the flag is not supplied.

Pushed

khyperia commented:
(For what it's worth, just confirming that the current code doesn't work with a full path - it surprisingly kind of almost does, but not quite, due to quirks of `GetProgramPath`)

cachemeifyoucan commented:
> I tried to take precedence from -fuse-ld and --ld-path, so -fuse-lipo takes a name, and a hypothetical future --lipo-path would be a full path. Perhaps these are only named this way due to legacy compatibility though, and new naming conventions should be something else, fuse-lipo-program as you say? I can't find any existing option that ends with -program=, though. I don't know the history and context here. Let me know what I should do!

Make sense to me.

khyperia commented:
> It is better to add a CLANG_DEFAULT_LIPO cmake option to match the CLANG_DEFAULT_LINKER for -fuse-ld.

Sure, that's a possibility, but in my opinion can be done in a future PR. There's a lot of additional features that can be done in the future:

- CLANG_DEFAULT_LIPO cmake option
- `--lipo-path` commandline argument
- `-fuse-llvm-darwin-tools`, described in the original issue
- better error handling when the lipo tool isn't found
- probably more...

I have a motivation and use case for `-fuse-lipo` and I'd like to get this in first, and we can talk about additional features later - there's no backcompat issues/etc. that require these additional features to be done within this PR.

khyperia commented:
@cachemeifyoucan Could you possibly help me get this merged? I'm not sure if your approval is enough, or if more reviewers are needed - and if so, who. Are you able to merge this if just your approval is enough?

khyperia commented:
@cachemeifyoucan thank you, done! (feel free to rewrite it as well, if you want or have any nitpicks)

cachemeifyoucan commented:
I updated the title and PR message. Let me know if that looks ok for you.

khyperia commented:
Looks great, thanks!

cachemeifyoucan commented:
Merged. Thanks for your contribution!
